ENH migrate GLMs / TweedieRegressor to linear loss

lorentzenchr opened the pull request:
#### Reference Issues/PRs
This is a follow-up of #21808 and #20567.
It also fixes #22124 (~partial fix of #21406).

#### What does this implement/fix? Explain your changes.
This PR plugs in the new `LinearModuleLoss` in the private`GeneralizedLinearRegressor`, thereby removing `sklearn._loss.glm_distribution.py` and `sklearn.linear_model._glm/link.py`.
The tweedie deviance code is copy&pasted into the metric `mean_tweedie_deviance`.

#### Any other comments?
It should be a user API backward compatible change (`PoissonRegressor`, `GammaRegressor` and `TweedieRegressor`, `mean_tweedie_deviance`).

lorentzenchr committed:
DOC fix missing power of 2

lorentzenchr committed:
FIX typo in error message

lorentzenchr committed:
FIX bug in loss tests

lorentzenchr committed:
FEA add HalfTweedieLossIdentity

lorentzenchr committed:
DOC comment about linear predictor

lorentzenchr committed:
ENH refactor glm.py with LinearModelLoss

lorentzenchr committed:
MAINT remove link.py in _glm submodule

lorentzenchr committed:
CLN googlemail to gmail

lorentzenchr committed:
MAINT copy tweedie deviance to regression metrics

lorentzenchr committed:
MAINT remove glm_distribution.py

lorentzenchr committed:
CLN fix comment

lorentzenchr committed:
CLN fix bug in tweedie deviance metrics

lorentzenchr committed:
FIX dtype because of lbfgs

lorentzenchr committed:
TST TweedieRegressor for validate in init

lorentzenchr committed:
ENH use _openmp_effective_n_threads

lorentzenchr committed:
DOC add whatsnew

lorentzenchr commented on sklearn/_loss/loss.py:
```diff
@@ -757,6 +758,51 @@ def constant_to_optimal_zero(self, y_true, sample_weight=None):
             return term
 
 
+class HalfTweedieLossIdentity(BaseLoss):
+    """Half Tweedie deviance loss with identity link, for regression.
```
This new loss class is needed for `TweedieRegressor(link="identity")`.
See Cython implementation in file `_loss.pyx.tp`.

lorentzenchr commented on sklearn/linear_model/_glm/glm.py:
```diff
@@ -79,19 +76,28 @@ class GeneralizedLinearRegressor(RegressorMixin, BaseEstimator):
+        base_loss_class          Link       Target Domain
+        =======================  ========  ==========================
+        HalfSquaredError         identity  y any real number
+        HalfPoissonLoss          log       0 <= y
+        HalfGammaLoss            log       0 < y
+        HalfInverseGaussianLoss  log       0 < y
+        HalfTweedieLoss          log       dependend on tweedie power
+        =======================  ========  ==========================
 
-    link : {'auto', 'identity', 'log'} or an instance of class BaseLink, \
-            default='auto'
         The link function of the GLM, i.e. mapping from linear predictor
-        `X @ coeff + intercept` to prediction `y_pred`. Option 'auto' sets
-        the link depending on the chosen family as follows:
+        `X @ coeff + intercept` to prediction `y_pred`. For instance, with a log link,
+        we have `y_pred = exp(X @ coeff + intercept)`.
 
-        - 'identity' for Normal distribution
-        - 'log' for Poisson,  Gamma and Inverse Gaussian distributions
+    base_loss_params : dictionary, default={}
+        Arguments to be passed to base_loss_class, e.g. {"power": 1.5} with
+        `base_loss_class=HalfTweedieLoss`.
```
`family` and `link` were private attributes. They are replaced by `base_loss_class` and `base_loss_params`. The new losses have the link functions baked into them.

lorentzenchr commented on sklearn/linear_model/_glm/glm.py:
```diff
@@ -269,57 +238,70 @@ def fit(self, X, y, sample_weight=None):
-        if not np.all(family.in_y_range(y)):
+        n_samples, n_features = X.shape
+
+        self._linear_loss = LinearModelLoss(
+            base_loss=self._get_base_loss_instance(),
+            fit_intercept=self.fit_intercept,
+        )
```
**This is the reason for this PR, to use `LinearModelLoss`**

lorentzenchr commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -441,47 +391,107 @@ def test_convergence_warning(regression_data):
+@pytest.mark.parametrize(
+    "name, link_class", [("identity", IdentityLink), ("log", LogLink)]
+)
+def test_tweedie_link_argument(name, link_class):
```
Replaces former `test_glm_link_argument`.

lorentzenchr commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -441,47 +391,107 @@ def test_convergence_warning(regression_data):
+        (3, LogLink),  # inverse-gaussian
+    ],
+)
+def test_tweedie_link_auto(power, expected_link_class):
```
Replaces former `test_glm_link_auto`

lorentzenchr commented on sklearn/metrics/_regression.py:
```diff
@@ -1026,8 +1028,56 @@ def mean_tweedie_deviance(y_true, y_pred, *, sample_weight=None, power=0):
+        target_type=numbers.Real,
+    )
+
+    message = f"Mean Tweedie deviance error with power={p} can only be used on "
```
From here on, it is really copy&paste from the now deleted `glm_distribution.py`.

lorentzenchr reviewed (commented):
Some hints for reviewers.

lorentzenchr committed:
FIX dtype in score funtion

lorentzenchr committed:
FIX validation of base_loss_class

lorentzenchr committed:
TST validation of y_true range

lorentzenchr commented:
This might interest @agramfort @rth @TomDLT.
After this, one could imagine to develop a better 2nd order optimizer.

lorentzenchr committed:
DOC fix typo

TomDLT commented on sklearn/linear_model/_glm/glm.py:
```diff
@@ -79,19 +76,28 @@ class GeneralizedLinearRegressor(RegressorMixin, BaseEstimator):
 
-        - 'identity' for Normal distribution
-        - 'log' for Poisson,  Gamma and Inverse Gaussian distributions
+    base_loss_params : dictionary, default={}
```
If `base_loss_class` is immediately instantiated with `base_loss_params`, would it be simpler to take a unique parameter `base_loss=base_loss_class(**base_loss_params)` ?

lorentzenchr replied:
See comment below, https://github.com/scikit-learn/scikit-learn/pull/22548#discussion_r811550321.

TomDLT commented on sklearn/linear_model/_glm/glm.py:
```diff
@@ -813,33 +815,48 @@ def __init__(
+        self.power = power
 
-    @property
-    def family(self):
```
How does a user know it was a private property? This change might need a deprecation cycle.

lorentzenchr replied:
You are right. But this would mean keeping the file `sklearn._loss.glm_distribution.py` because, e.g.,
```python
mod = TweedieRegressor(power=1.5)
mod.family
```
gives `sklearn._loss.glm_distribution.TweedieDistribution`. We should have made it private from the beginning. Note that it only appears in the API doc because it is inherited (from a private parent).

Proposition:
- Let's deprecate `family` and always return a DeprecationWarning.
- Keep current behavor for `PoissonRegressor` and `GammaRegressor` (returning strings  `"poisson"` and `"gamma"`, respectively).
- Breaking change for `TweedieRegressor`, return `"tweedie"` (or `None` or ...)

lorentzenchr replied:
See https://github.com/scikit-learn/scikit-learn/pull/22548/commits/db6c1928a22a1d6c82986b03ab9219e5f1f57e20

TomDLT commented on sklearn/linear_model/_glm/glm.py:
```diff
@@ -136,8 +142,8 @@ def __init__(
         fit_intercept=True,
-        family="normal",
-        link="auto",
+        base_loss_class=HalfSquaredError,
```
If these parameters are private, should we name them with a leading underscore `_base_loss_class` ?

TomDLT commented on sklearn/linear_model/_glm/glm.py:
```diff
@@ -661,28 +670,20 @@ def __init__(
-        if value != "gamma":
-            raise ValueError("GammaRegressor.family must be 'gamma'!")
+    def _get_base_loss_instance(self):
+        if self.base_loss_class != HalfGammaLoss:
```
Is it useful to check `self.base_loss_class` if it is never used ? It seems like `self.base_loss_class` is hardcoded anyway, so I don't understand the goal of this pattern.

TomDLT replied:
I find the `_get_base_loss_instance` / `self.base_loss_class` / `self.base_loss_params` pattern confusing. In all public estimators, the parameters are hardcoded, checked, but actually not used.

 I think we should either:
- use a unique private parameter `_base_loss`, hardcoded when the public estimator is instantiated.
- use an abstract method `_get_base_loss`, overwritten by each public estimator.

lorentzenchr replied:
I'd prefer the first option. The presence of `link`, however, makes it more complicated. We would like to call `super().__init__()` inside `__init__`. There, we would have to translate the `link` argument into the right loss class. This would mean checking the validity of `link` inside `__init__` which is an antipattern for scikit-learn.

Therefore, I'll go (for the moment) with option 2: abstract method `_get_base_loss`.

lorentzenchr replied:
But removing `base_loss_params` is also a little unsatisfactory. Without it, we can't instantiate a `_GeneralizedLinearRegressor` with a loss with parameter like `HalfTweedieLoss(power=1.5)`.

lorentzenchr replied:
I removed `base_loss_params` in a0be233f6a5dfb8e7efe71d90ae7dfd87d294f87. Let me know if you like it more.

With a0be233f6a5dfb8e7efe71d90ae7dfd87d294f87, one sets a class instance to `_base_loss`. This is validated with `_validate_base_loss`. In `TweedieRegressor`, `_base_loss` is neglected and `_validate_base_loss` returns the used loss during `fit`.

TomDLT replied:
>I think we should either:
>   - use a unique private parameter `_base_loss`, hardcoded when the public estimator is instantiated.
>   - use an abstract method `_get_base_loss`, overwritten by each public estimator.

It seems that the current solution use **both** options. It uses a hard-coded private parameter **and** a method `_validate_base_loss` overwritten by each public estimator. I think it is weird to validate a hard-coded parameter. I understand we cannot only have a private parameter, in order to allow `TweedieRegressor` to specify the power and link.

I think a simpler pattern would be to remove the hardcoded private parameter `_base_loss`, and to only use a method `_get_base_loss` overwritten by each public estimator.

thomasjpfan replied:
I opened a small PR to @lorentzenchr's fork to remove the hardcoded `_base_loss`: https://github.com/lorentzenchr/scikit-learn/pull/3

lorentzenchr replied:
You are right. @thomasjpfan opened https://github.com/lorentzenchr/scikit-learn/pull/3 with your (re)solution. I'll merge that.

TomDLT commented on sklearn/linear_model/_glm/glm.py:
```diff
@@ -731,10 +732,11 @@ class TweedieRegressor(GeneralizedLinearRegressor):
         The link function of the GLM, i.e. mapping from linear predictor
         `X @ coeff + intercept` to prediction `y_pred`. Option 'auto' sets
-        the link depending on the chosen family as follows:
+        the link depending on the chosen base_loss_class (EDM family) as follows:
```
I don't think we should mention the private parameter `base_loss_class`.

TomDLT commented on sklearn/linear_model/_glm/glm.py:
```diff
@@ -813,33 +815,48 @@ def __init__(
-            family=TweedieDistribution(power=power),
-            link=link,
+            base_loss_class=HalfTweedieLoss,
+            base_loss_params={"power": power},
```
Duplicating the path for `power` might lead to bugs. I would not put it in `base_loss_params`, since we use `self.base_loss_params["power"] = self.power` later anyway.

lorentzenchr replied:
Resolved by https://github.com/lorentzenchr/scikit-learn/pull/3.

lorentzenchr committed:
MNT make GeneralizedLinearRegressor private

lorentzenchr commented on sklearn/linear_model/_glm/glm.py:
```diff
@@ -79,19 +76,28 @@ class GeneralizedLinearRegressor(RegressorMixin, BaseEstimator):
-            or an ExponentialDispersionModel instance, default='normal'
-        The distributional assumption of the GLM, i.e. which distribution from
-        the EDM, specifies the loss function to be minimized.
+    base_loss_class : subclass of BaseLoss, default=HalfSquaredError
```
I made in private by the convention of prefixing with `_` in 3b6beda8b490f6547619a009d576d60315e4050f.

lorentzenchr committed:
DOC typo and text improvements

lorentzenchr committed:
MNT make base_loss_class private

jjerphan commented on doc/modules/linear_model.rst:
```diff
@@ -1032,7 +1032,7 @@ reproductive exponential dispersion model (EDM) [11]_).
 The minimization problem becomes:
 
-.. math::    \min_{w} \frac{1}{2 n_{\text{samples}}} \sum_i d(y_i, \hat{y}_i) + \frac{\alpha}{2} ||w||_2,
+.. math::    \min_{w} \frac{1}{2 n_{\text{samples}}} \sum_i d(y_i, \hat{y}_i) + \frac{\alpha}{2} ||w||_2^2,
```
Nice catch.

lorentzenchr replied:
thanks

jjerphan commented on sklearn/_loss/loss.py:
```diff
@@ -757,6 +758,51 @@ def constant_to_optimal_zero(self, y_true, sample_weight=None):
+                    - y_true_i * raw_prediction_i**(1-p) / (1-p)
+                    + raw_prediction_i**(2-p) / (2-p)
+
+    Note that the minimum loss is 0.
```
```suggestion
    Note that the minimum value for this loss is 0.
```

jjerphan reviewed (commented):
First pass. The Cython implementation LGTM.

lorentzenchr commented on sklearn/_loss/_loss.pyx.tp:
```diff
@@ -501,6 +529,104 @@ cdef inline double_pair cgrad_hess_half_tweedie(
+        return tmp
+
+
+cdef inline double cgradient_half_tweedie_identity(
+    double y_true,
+    double raw_prediction,
+    double power
+) nogil:
+    cdef double tmp
+    if power == 0.:
+        return raw_prediction - y_true
+    elif power == 1.:
+        return 1. - y_true / raw_prediction
+    elif power == 2.:
+        return (raw_prediction - y_true) / (raw_prediction * raw_prediction)
+    else:
+        tmp = pow(raw_prediction, -power)
+        return tmp * (raw_prediction - y_true)
```
removed `tmp`

lorentzenchr committed:
CLN address review comments

lorentzenchr committed:
Merge branch 'main' into migrate_glm_to_linear_loss

ogrisel committed:
More valid cases for HalfTweedieLossIdentity(power=0)

ogrisel reviewed (commented):
I am currently doing a pass of review on this PR. But before completing it would like to already share that I observed a speed-up of 2x on a typical Poisson regression problem derived from our example.

Also, memory usage remained the same and there is no sign of any memory leak.

Here is the benchmark script I used:

<details>

```python
from time import perf_counter
import numpy as np
from sklearn.datasets import fetch_openml
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import FunctionTransformer, OneHotEncoder
from sklearn.preprocessing import StandardScaler, KBinsDiscretizer
from sklearn.compose import ColumnTransformer
from sklearn.linear_model import PoissonRegressor
from sklearn.metrics import mean_poisson_deviance
from sklearn.model_selection import train_test_split
from joblib import Memory


m = Memory(location="/tmp/cache")


log_scale_transformer = make_pipeline(
    FunctionTransformer(np.log, validate=False), StandardScaler()
)
linear_model_preprocessor = ColumnTransformer(
    [
        ("passthrough_numeric", "passthrough", ["BonusMalus"]),
        (
            "binned_numeric",
            KBinsDiscretizer(n_bins=10, subsample=int(1e5), random_state=0),
            ["VehAge", "DrivAge"],
        ),
        ("log_scaled_numeric", log_scale_transformer, ["Density"]),
        (
            "onehot_categorical",
            OneHotEncoder(),
            ["VehBrand", "VehPower", "VehGas", "Region", "Area"],
        ),
    ],
    remainder="drop",
)


@m.cache
def load_preprocess_data(test_size=0.33, random_state=0):
    df = fetch_openml(data_id=41214, as_frame=True).frame
    df["Frequency"] = df["ClaimNb"] / df["Exposure"]
    df_train, df_test = train_test_split(df, test_size=test_size, random_state=random_state)

    df_train_transformed = linear_model_preprocessor.fit_transform(df_train)
    df_test_transformed = linear_model_preprocessor.transform(df_test)

    return df_train, df_test, df_train_transformed, df_test_transformed


print("Loading data...")
df_train, df_test, df_train_transformed, df_test_transformed = load_preprocess_data()

durations = []
for i in range(5):
    print("Training model...")
    poisson_glm = PoissonRegressor(alpha=1e-12, max_iter=300)
    t0 = perf_counter()
    poisson_glm.fit(
        df_train_transformed, df_train["Frequency"], sample_weight=df_train["Exposure"]
    )
    t1 = perf_counter()
    duration = (t1 - t0)
    durations.append(duration)
    print(f"Training time: {duration:.3f}s")
    test_poisson_deviance = mean_poisson_deviance(
        df_test["Frequency"],
        poisson_glm.predict(df_test_transformed),
        sample_weight=df_test["Exposure"],
    )
    print(f"Test deviance: {test_poisson_deviance:.3f}")

print(f"Average training time: {np.mean(durations):.3f}s")

```

</details>

and the results:

- on main:

```
Loading data...
Training model...
Training time: 3.872s
Test deviance: 0.594
Training model...
Training time: 3.916s
Test deviance: 0.594
Training model...
Training time: 3.893s
Test deviance: 0.594
Training model...
Training time: 3.868s
Test deviance: 0.594
Training model...
Training time: 3.891s
Test deviance: 0.594
Average training time: 3.888s
```

![poisson_insurance_main](https://user-images.githubusercontent.com/89061/157419725-92e172e0-04bb-4f3f-b707-1c36526e6a58.png)


- on this branch:

```
Loading data...
Training model...
Training time: 1.826s
Test deviance: 0.594
Training model...
Training time: 1.847s
Test deviance: 0.594
Training model...
Training time: 1.802s
Test deviance: 0.594
Training model...
Training time: 1.813s
Test deviance: 0.594
Training model...
Training time: 1.805s
Test deviance: 0.594
Average training time: 1.818s
```

![poisson_insurance_branch](https://user-images.githubusercontent.com/89061/157419749-8f31034c-6a92-4be7-a86a-f3ff00b99330.png)

Since github does not allow me to stash my pending review notes to post this comment, please find below some cosmetic notes I took on the way. More important comments will follow.

ogrisel commented:
Here are the viztracer traces of a single poisson fit (you can see the preprocessed data loaded from a joblib cache) and a zoom on a single LBFGS iteration.

![Screenshot 2022-03-09 at 11 16 37](https://user-images.githubusercontent.com/89061/157422374-2026cb54-52c3-4fd4-b3f0-bde9844be30a.png)

![Screenshot 2022-03-09 at 11 18 53](https://user-images.githubusercontent.com/89061/157422365-cf39749a-6fb7-48a8-b7c5-adb73380192a.png)

Nothing fishy. Contrary to logistic regression, the here the computation of the gradient of the loss w.r.t. the inputs of the loss function are not negligible compared to the dot products that involve `X_train`. For information `X_train` has shape `(454268, 75)` using this feature preprocessing.

ogrisel committed:
Add test_tweedie_log_identity_consistency

ogrisel reviewed (commented):
Another pass of comments with a new test I just pushed to evaluate the consistency of the new loss class introduced in this PR. I haven't yet completed my review but have to stop here for today.

ogrisel commented on sklearn/linear_model/_glm/glm.py:
```diff
@@ -813,33 +814,48 @@ def __init__(
-                "TweedieRegressor.family must be of type TweedieDistribution!"
+            raise ValueError(
+                "The link must be an element of ['auto', 'identity', 'log']; "
+                "got (link={0})".format(self.link)
```
```suggestion
                f"got (link={self.link!r})"
```

ogrisel committed:
Did not mean to use a logspace for raw_prediction

ogrisel commented on sklearn/_loss/tests/test_loss.py:
```diff
@@ -1091,3 +1126,50 @@ def test_loss_pickle(loss):
+    # XXX: increasing the upper bound makes the test fail very easily, first
+    # with significantly different gradients, and then with different loss
+    # values.
+    raw_prediction = np.logspace(-10, -3, n_samples)
```
I did not mean to use `np.logspace` but a `np.linspace` here. Otherwise taking `np.exp(raw_prediction)` would be very large. Fixed in 80bf4702223f624ffcd66625ed8f76cb966a8c08.

ogrisel commented on sklearn/_loss/tests/test_loss.py:
```diff
@@ -1091,3 +1126,50 @@ def test_loss_pickle(loss):
+    assert_allclose(gradient_log, gradient_identity, atol=1e-3)
+
+    # XXX: the following always fails
+    # assert_allclose(hessian_log, hessian_identity, atol=1e-3)
```
I updated the test in [80bf470](https://github.com/scikit-learn/scikit-learn/pull/22548/commits/80bf4702223f624ffcd66625ed8f76cb966a8c08) but the gradient and hessian computation still do not match as I would have expected. Maybe I am making a mistake in the test. I have not yet checked the source of that function to understand the discrepancy.

lorentzenchr replied:
It's because of the chain rule with the (log) link function, e.g. `d/dx loss(exp(x)) = exp(x) * grad(exp(x))` and so on.

The two losses must be equal if we add `constant_to_optimal_zero`, no need for `diff`.

Furthermore, we only need to check the losses. Gradients and hessians are tested in `test_gradients_hessians_numerically` (in a way that I can sleep well).

lorentzenchr replied:
Fixed in https://github.com/scikit-learn/scikit-learn/pull/22548/commits/225863b4d82d2f0d164fae1142ae3009a60ba295.

lorentzenchr committed:
CLN address review comments

lorentzenchr committed:
DEP family
- PoissonRegressor with usual deprecration
- GammaRegressor with usual deprecation
- TweedieRegressor breaking deprecation
  Reason: We want to get rid of the `class ExponentialDispersionModel` entirely

lorentzenchr committed:
CLN remove base_loss_params

lorentzenchr committed:
TST fix test_tweedie_log_identity_consistency

ogrisel commented:
> Following the discussion https://github.com/scikit-learn/scikit-learn/pull/22548#discussion_r811542255, I set a proper deprecation for the family attribute in https://github.com/scikit-learn/scikit-learn/commit/db6c1928a22a1d6c82986b03ab9219e5f1f57e20. For PoissonRegressor and GammaRegressor this is fine as familiy is a string. For TweedieRegressor this is a sort of breaking change, as we want to get rid of the class ExponentialDispersionModel entirely.
Note that we made a mistake by not marking family nor GeneralizedLinearRegressor as private by prefixing with an underscore. family then appears in the doc of, e.g., PoissonRegressor as it is inherited.

Maybe this is an acceptable breaking change since it's the only sane way to fix the `__init__` param validation common test but it should be documented as such in the changelog.

I still need to find the time to conduct a full review but I based on my initial review, this PR seems strong so don't wait for me if two other reviewers give their +1 in the mean time.

lorentzenchr committed:
DOC add deprecation and change info in whatsnew

ogrisel commented on sklearn/linear_model/_glm/glm.py:
```diff
@@ -80,18 +77,33 @@ class GeneralizedLinearRegressor(RegressorMixin, BaseEstimator):
-        the link depending on the chosen family as follows:
+        .. deprecated:: 1.1
+           `family` is deprecated in 1.1 and will be removed in 1.3.
+           Use `_base_loss` instead.
```
```suggestion
           Use the private `_base_loss` instead.
```

ogrisel replied:
Actually I am not sure we need a deprecation warning here since the `GeneralizedLinearRegressor` class is private (it was not exposed under `sklearn.linear_model`) and the family constructor parameter is not exposed in the public concrete descendant classes.

ogrisel commented on sklearn/linear_model/_glm/glm.py:
```diff
@@ -423,27 +404,81 @@ def score(self, X, y, sample_weight=None):
         else:
-            raise ValueError
-        return {"requires_positive_y": not _family_instance.in_y_range(-1.0)}
+            return self._base_loss.__class__.__name__
```
I don't like deprecating without preserving the previous behavior. Either we delete this attribute entirely and document it as a breaking change or we keep it with the previous return values (not strings that are useless).

I have the feeling that it would be simpler to keep the `sklearn/_loss/_glm_distribution.py` file around during the deprecation cycle of the family attribute. We can just put a header at the top of the source file to make it explicit that this file is kept around until 1.3 just to preserve backward compat for the deprecated `family` parameter.

ogrisel reviewed (commented):
I think we should do a regular deprecation cycle for the family parameter. I know this impose keeping around an other wise useless Python module for 2 releases, but so be it.

ogrisel commented on doc/whats_new/v1.1.rst:
```diff
@@ -559,6 +565,14 @@ Changelog
   :user:`Olivier Grisel <ogrisel>`.
 
+- |Fix| The input parameter `family` of :class:`linear_model.TweedieRegressor` is not
+  validated in `__init__` anymore. Instead, this (private) parameter is deprecated in
```
Actually the public `TweedieRegression` class had no input parameter named `family`, only an instance attribute named `family` (which is now deprecated).

lorentzenchr commented:
> I think we should do a regular deprecation cycle for the family parameter. I know this impose keeping around an other wise useless Python module for 2 releases, but so be it.

What changed your mind?
I would like to hear a second reviewer's opinion to avoid back and forth, even if it is almost as simple as undo 719bb1be3ee049be23ae927efb357578d04949f7.

lorentzenchr committed:
Revert "MAINT remove glm_distribution.py"

This reverts commit 719bb1be3ee049be23ae927efb357578d04949f7.

lorentzenchr committed:
MNT add removal TODO for glm_distribution.py

lorentzenchr committed:
DEP backward compatible deprecation of family property

lorentzenchr commented:
> > I think we should do a regular deprecation cycle for the family parameter. I know this impose keeping around an other wise useless Python module for 2 releases, but so be it.
> 
> What changed your mind? I would like to hear a second reviewer's opinion to avoid back and forth, even if it is almost as simple as undo [719bb1b](https://github.com/scikit-learn/scikit-learn/commit/719bb1be3ee049be23ae927efb357578d04949f7).

I changed my mind, too. Let's do proper deprecation and hopefully move forward with this. This PR is the last missing peace for possible new 2nd order solvers for all/most GLMs!

lorentzenchr committed:
Merge branch 'main' into migrate_glm_to_linear_loss

thomasjpfan committed:
CLN Idea removing _base_loss

ogrisel reviewed (approved):
LGTM, thank you very much for the PR.

thomasjpfan reviewed (commented):
I opened a PR to your fork to showcase a way to avoid `_base_loss`:
https://github.com/lorentzenchr/scikit-learn/pull/3

lorentzenchr committed:
Merge pull request #3 from thomasjpfan/pr/22548_idea

CLN Idea removing _base_loss

lorentzenchr committed:
CLN fix type in test name

thomasjpfan commented on sklearn/metrics/_regression.py:
```diff
@@ -1182,24 +1232,20 @@ def d2_tweedie_score(y_true, y_pred, *, sample_weight=None, power=0):
-    y_avg = np.average(y_true, weights=sample_weight)
-    dev = dist.unit_deviance(y_true, y_avg, check_input=True)
-    denominator = np.average(dev, weights=sample_weight)
+    y_avg = np.full_like(y_true, np.average(y_true, weights=sample_weight))
```
I think this is a memory regression. Previously, `y_avg` was a single value and `dist.unit_deviance` used broadcasting to compute the denominator. Now `y_avg` needs to be a full array to be passed into `mean_tweedie_deviance`.

Without changing the public API of `mean_tweedie_deviance`, a workaround is to define a `_mean_tweedie_deviance` that does not require consistent lengths and call it here and in `mean_tweedie_deviance`. Something like this: https://github.com/lorentzenchr/scikit-learn/pull/4

lorentzenchr replied:
Thanks for the PR!

thomasjpfan committed:
Small refactor for _mean_tweedie_deviance

thomasjpfan reviewed (commented):
I doubled checked the math for `HalfTweedieLossIdentity` and it looks good to me. I have a few small comments on the code.

There is an issue with `d2_tweedie_score` in terms of memory.

lorentzenchr committed:
Merge pull request #4 from thomasjpfan/pr/22548_memory

Small refactor for _mean_tweedie_deviance

lorentzenchr committed:
CLN fix merge conflict in whatsnew

lorentzenchr committed:
Merge branch 'migrate_glm_to_linear_loss' of https://github.com/lorentzenchr/scikit-learn into migrate_glm_to_linear_loss

lorentzenchr commented on sklearn/linear_model/_glm/glm.py:
```diff
@@ -11,57 +11,54 @@
-        return X @ coef[1:] + coef[0]
-    else:
-        return X @ coef
+class HalfInverseGaussianLoss(HalfTweedieLoss):
```
No. Not with the new `_get_loss` logic.

lorentzenchr committed:
CLN remove HalfInverseGaussianLoss

lorentzenchr committed:
DOC move private _base_loss to attributes section

lorentzenchr committed:
CLN remove comment

lorentzenchr committed:
CLN TODO(1.3)

thomasjpfan commented on sklearn/_loss/loss.py:
```diff
@@ -770,6 +771,52 @@ def constant_to_optimal_zero(self, y_true, sample_weight=None):
             return term
 
 
+class HalfTweedieLossIdentity(BaseLoss):
```
My curiosity: when does it make sense to use identity link with power != 0 ?

lorentzenchr replied:
Years ago, I thought it a good idea. Meanwhile, I don't think it's is useful. Therefore, I opened #19086 without much response.

jpbrodsky replied:
> My curiosity: when does it make sense to use identity link with power != 0 ?

I have a problem where I know the expectation y' follows the linear model y' = w x. My measurements, y, have poisson errors. 

(The specific problem involves analysis of radiation measurements. The expectation is linear with the amount of source; the measurements are poisson distributed).

Using a log link function is just not the right description of my problem. Yes, the whole thing breaks down when evaluating negative values of w, but it seems much better to offer a constraint to avoid ever evaluating negative values of w rather than exclude the situations where you have an actual linear relationship with poisson measurements.

thomasjpfan reviewed (approved):
I ran the benchmarks from https://github.com/scikit-learn/scikit-learn/pull/22548#pullrequestreview-903395828 and see the same performance improvements.

LGTM
